### PR TITLE
[Support] Add `\g<ref>` backreferences in Regex::sub()

### PR DESCRIPTION
**Pull request by igorkudrin**
The existing format of backreferences, `\<ref>`, does not allow digits to be placed directly after the reference because they are included in the reference number. The new format solves this problem by adding explicit delimiters.

### CODE REVIEW

**Comment by dwblaikie** [comment]
I guess a workaround would be to add some empty braces to separate the matching number from the following digits? (`\3()42`)? Maybe that'd be sufficient, if a bit quirky? (I'm not totally opposed to this patch/feature addition either)

How's this compare to other regex libraries & what syntax they use for this situation?

**Comment by igorkudrin** [comment]
> I guess a workaround would be to add some empty braces to separate the matching number from the following digits? (`\3()42`)? Maybe that'd be sufficient, if a bit quirky? (I'm not totally opposed to this patch/feature addition either)

The problem is not about regexes, but substitutions, which have their own syntax. Basically, they copy the template into the output string, replacing sequences like "\\n" and "\\t" with '\n' and '\t', and substituting '\\' followed by digits with the found groups from the matched regex. This syntax has no its own groups, so adding "()" would just copy these characters to the output.

> How's this compare to other regex libraries & what syntax they use for this situation?

- [Python](https://docs.python.org/3/library/re.html#re.sub): "\\g<num>"
- [.Net](https://learn.microsoft.com/en-us/dotnet/standard/base-types/substitutions-in-regular-expressions#substituting-a-numbered-group): "${num}"
- [ECMAScript](https://tc39.es/ecma262/multipage/text-processing.html#sec-string.prototype.replace) doesn't seem to have a special syntax for group numbers, but has "$<name>" to substitute a named group, which can be used instead.

**Comment by dwblaikie** [comment]
Worth using a syntax that matches something existing, like the Python one seems close? Or do we already diverge in a bunch of ways for the syntax in substitutions that there's no real consistency to be gained?

**Comment by igorkudrin** [comment]
> Worth using a syntax that matches something existing, like the Python one seems close? Or do we already diverge in a bunch of ways for the syntax in substitutions that there's no real consistency to be gained?

I'm not sure what syntax the original author was trying to mimic, although Python's looks very close. I didn't go this way because supporting the `\\g<num>` notation requires a bit more code and, as you can see, each implementation uses its own variant anyway. This particular syntax from Python looks very specific, so I supposed that implementing it might mislead a developer into thinking that other features are also supported; I preferred to stick with something that looks more general. Anyway, I have no objection to implementing any other variant as long as it solves the problem. What is your preference?

**Comment by dwblaikie** [comment]
Yeah, looks like Python's path probably started somewhere similar to us. Let's go with the python syntax, if you could - thanks!

File: llvm/unittests/Support/RegexTest.cpp [hunk 1]
```diff
@@ -127,6 +127,22 @@ TEST_F(RegexTest, Substitution) {
...
+  EXPECT_EQ("0g<15", Regex("a([0-9]+).*").sub("0\\g<15", "a1234ber", &Error));
+  EXPECT_EQ("", Error);
+
+  EXPECT_EQ("0g<>15", Regex("a([0-9]+).*").sub("0\\g<>15", "a1234ber", &Error));
+  EXPECT_EQ("", Error);
+
+  EXPECT_EQ("0g<3e>1",
+            Regex("a([0-9]+).*").sub("0\\g<3e>1", "a1234ber", &Error));
+  EXPECT_EQ("", Error);
+
```
Comment (MaskRay) [inline llvm/unittests/Support/RegexTest.cpp:143]: Add a test demonstrating `\\g<10>` referencing the 10th capture group.

**Comment by MaskRay** [comment]
Using `\g` for replacement looks good. https://www.regular-expressions.info/replacebackref.html

**Comment by igorkudrin** [comment]
> Using `\g` for replacement looks good. https://www.regular-expressions.info/replacebackref.html

Thanks for the link. It's interesting that they mention the `\g<ref>` syntax only for referencing named groups, which we do not support.

**Review by dwblaikie — APPROVED** [review]

**Pull request merged**